### PR TITLE
Delaunay triangulation and Voronoi diagrams

### PR DESCRIPTION
**Pull request by folkertdev**
remaining issues: 

delaunay:
* speed up finding close triangles using a spatial tree 
* speed up new face creation and insertion

based on the data from the chrome devtools, these operations take roughly equal amounts of time at the moment. 

voronoi:
* no profiling has been done for voronoi yet.
* how do we deal with the boundaries of the diagram. See the image where for instance in the lower left there is no polygon. Likely this means specifying some viewport, so we can cull polygons that go toward infinity at the edges of the viewport. The faces that go toward infinity are likely based on the `SuperTrianglePoints`. 

voronoi-border

### CODE REVIEW

**Comment by folkertdev** [comment]
commit messages speak for themselves. I've also been playing with [this paper on bounding the voronoi polygons](https://www.isprs-ann-photogramm-remote-sens-spatial-inf-sci.net/II-2-W2/235/2015/isprsannals-II-2-W2-235-2015.pdf). Not the best solution for all cases but the example they show is convincing.

**Comment by ianmackenzie** [comment]
One thought on the Voronoi diagram: maybe don't *try* to bound the regions, just represent them as something like

```elm
type VoronoiRegion
    = Bounded Polygon2d
    | Unbounded Polyline2d Direction2d Direction2d
```
where in the `Unbounded` case the polyline represents the actual boundary lines, and the two directions represent the directions of the two lines extending off to infinity from the two ends of the polyline. Then callers could decide what to do...and we could still provide separate functions for taking a `List VoronoiRegion` and trimming them to a bounding box to get a `List Polygon2d`. Thoughts?

**Comment by folkertdev** [comment]
Sounds good, though it doesn't really make the problem easier. How can we find these directions? 

I've experimented with including or otherwise treating the points at infinity in a clever way, but I'm not quite sure what I'm looking for. In this more realistic data set (from the paper I linked, it's dialects of dutch spoken in the Netherlands and Belgium).

with-zero

in the north and west there are some data points that aren't in any polygon. The image above only includes bounded polygons (so, 0 points shared with the supertriangle). Now if I include the centerpoint of triangles with one point point at infinity, and that point is the first point (at (-inf, -inf)) I get an almost correct image: 

with-one-zero

But there is one point added that causes overlaps (incompetently circled in black). including the other two points at infinity, or allowing faces with 2 points at infinity, causes chaos. On the other hand, picking the 0th vertex and only include that one seems arbitrary and it might just be this dataset that makes it work. I've been trying to find information on how the bounding box works with bowyer-watson but can't find anything helpful on it. 

Another thing, based on #38: should the triangulation and diagram be opaque types, so we can extend them after construction? at least for delaunay that seems easily possible.

**Comment by folkertdev** [comment]
progress update: this now works 

success

I'm currently at 
```elm
type VoronoiRegion
    = Bounded Polygon2d
    | Unbounded (List Point2d) Direction2d Direction2d
```
The points are not in order, so a `Polyline2d` doesn't work for now. I'm don't know whether generating the points in order is possible (or that it is faster than using `convexHull`). 

Then this is what I use to render the polygons correctly: 
```elm
makeFinite direction =
    Point2d.origin |> Point2d.translateIn direction 2000000
```
There has to be some way around using a very large number there, but haven't figured that out yet.

**Comment by folkertdev** [comment]
I have a promising solution for generating the outside polygons without using very large numbers. 

[This page shows the problem](https://folkertdev.github.io/delaunay-voronoi/points-at-infinity.html) (move the slider). 

A point at infinity starts at the center of the edge between the other two points of its triangle, and 
travels in the direction perpendicular to that edge. 

When the edge is very long, it can be that other voronoi regions overlap with the center of that edge. 
In this case the point at infinity will create overlapping, wrong lines. The solution I have is to move the point at infinity at least far enough that it is outside of the bounding box of the remainder of the region. 
(i.e. it calculates the bounding box, and moves the point by the length of the diagonal of the bounding box). This method seems robust and cheap.

**Comment by ianmackenzie** [comment]
While I was looking through the Delaunay triangulation code it occurred to me that it should be possible to avoid tracking next/previous edges and coedges etc. by index...I've just added an experimental alternate implementation that does that. Seems promising so far (no array lookups at all!) but I'll have to do some more testing/profiling to see. And I still have to see if there's a good way to extract the Voronoi tesselation when using this approach...I think I have a method that will work but I haven't actually tried to implement it yet.

**Comment by ianmackenzie** [comment]
From a discussion with @folkertdev on Slack: calculating pseudo-angles could be a good way to perform Voronoi diagram generation (find all points on the Voronoi polygon for each vertex, then sort by pseudo-angle around that vertex to put the points in the correct order). Some good pseudo-angle algorithms: https://stackoverflow.com/questions/16542042/fastest-way-to-sort-vectors-by-angle-without-actually-computing-that-angle/16561333

**Comment by folkertdev** [comment]
Really nice work, apparently 3x faster than my approach already. It looks like you can also prevent two intermediate lists in `triangulation`. The first one (`Array.toList`) is an easy win, the second one (preventing `Dict.values`) gives a bit worse code but seems to help a little bit. 

From what I can see most remaining time is spent on Dict.foldl. It seems impossible to peak into the folded function to see what is actually spending time.  

Also I wonder whether we can use the idea that, because the points are sorted, it is quite likely that the triangle that was most-recently added will be removed by the following point. 

```elm
triangulation : Array Point2d -> List Face
triangulation points =
    let
        vertices =
            points |> Array.indexedMap Vertex

        numVertices =
            Array.length vertices

        uniqueVertices =
            Array.foldl insertVertex Dict.empty vertices

        topOuterIndex =
            numVertices

        leftOuterIndex =
            numVertices + 1

        rightOuterIndex =
            numVertices + 2

        totalNumVertices =
            numVertices + 3

        createInitialFaces firstVertex =
            [ OneVertexFace firstVertex
                topOuterIndex
                leftOuterIndex
                Direction2d.positiveY
            , OneVertexFace firstVertex
                leftOuterIndex
                rightOuterIndex
                Direction2d.negativeX
            , OneVertexFace firstVertex
                rightOuterIndex
                topOuterIndex
                Direction2d.negativeY
            ]

        folder _ vertex accumulator =
            case accumulator of
                [] ->
                    createInitialFaces vertex

                faces ->
                    let
                        ( retainedFaces, starEdges ) =
                            processFaces faces numVertices vertex [] Dict.empty

                        updatedFaces =
                            Dict.foldl (addNewFace vertex) retainedFaces starEdges
                    in
                        updatedFaces
    in
        Dict.foldl folder [] uniqueVertices
```

**Comment by ianmackenzie** [comment]
Interesting thought on insertion order - although I would like to try random insertion order to help with R-tree construction. And good point on the initial `toList`! I think it should be possible to make that even faster by converting to a list, sorting and then removing successive duplicates, though - sorting should be a lot faster than using a `Dict` for deduplication, that bit of code I threw together fairly quickly to get things working. Worth trying both approaches though!

**Comment by ianmackenzie** [comment]
I think that first image looks fine - the dashed lines are just drawn with a length of 700 pixels right now, there's no actual proper clipping logic attempted. So yes, if the start point of the infinite edge is far enough outside the drawing area, the line won't show up at all. What I put in there for now wasn't intended to be a full clipping solution, just a visualization to make sure the algorithm was working properly. Is there some error I'm missing, other than the length of the dashed lines?

Not sure about the R-tree! I'd hope that with randomized input order you should be able to get down to O(n log n) from O(n^2) but the constant factors might be higher due to the increased complexity of the code. There's a [good article from Mapbox](https://blog.mapbox.com/a-dive-into-spatial-search-algorithms-ebd0c5e39d2a) on the general principles of R-trees, but it doesn't go into great detail. If it helps, I've implemented R-tree variants myself in [Scala](https://github.com/ianmackenzie/opensolid-geometry-scala/blob/master/src/main/scala/org/opensolid/core/SpatialTree.scala) and [C++](https://github.com/ianmackenzie/opensolid-geometry-cpp/blob/master/OpenSolid/Core/SpatialSet.hpp), but I can't say they're fully commented!

OK, I'm getting fairly convinced that custom `DelaunayTriangulation` and `VoronoiDiagram` return types are probably the right approach. (That real time update example is really cool!) Thoughts on an API for those? I might try to take a stab at sketching out an API tonight or this weekend...

Side note: unless I'm missing something, using a Voronoi diagram for the purpose of "highlight the point nearest the mouse" is really silly. To do that efficiently, you'd have to have an R-tree of the Voronoi regions. And if you can use R-trees, then it would be simpler and more efficient to just to build an R-tree of the original points, since an R-tree itself can be used to do efficient "nearest point" queries (the algorithm for doing so is described in [the same Mapbox article](https://blog.mapbox.com/a-dive-into-spatial-search-algorithms-ebd0c5e39d2a) mentioned earlier).

Is it really true that `processFaces` is dominated by insertions? I guess if a newly inserted point is only in one face's circumcircle, then all that will end up happening is that the corresponding face's three edges will get inserted into `edgesByKey`. But if there are several faces that need to be removed, then *most* edges that get inserted into `edgesByKey` will be later removed when their coedge is added later. And I'd be reluctant to add a new dependency if the speedup was only minor anyways...

**Comment by folkertdev** [comment]
The problem is just with rendering I think. 

For `processFaces` most of the time a small number of faces is removed. For instance given a square with a diagonal, then there are 5 inserts and 1 removal. I agree that the difference is too small to justify the dependency, but in the future we might get some improvements because the dict implementation becomes faster (potentially the hashdict in 0.19 (if that happens) is also better). 

Something else to think about: rtrees that are fast for search, insert and delete. It looks like your previous implementations don't do inserts/deletes (or maybe my c++/scala reading skills are worse than I thought and I missed them).  

I experimented with this structure, based on 

https://github.com/aluuu/spatial_index/blob/master/src/rtree.ml

```elm
type RTree a
    = Empty
    | NonEmpty (NonEmptyRTree a)


type NonEmptyRTree a
    = Leaf { boundingBox : BoundingBox2d, values : List ( BoundingBox2d, a ) }
    | Node { boundingBox : BoundingBox2d, children : ( NonEmptyRTree a, List (NonEmptyRTree a) ) }
```

Removing the `Empty` case is great, and I also like the (a, List a) non-empty list construction for children, but it looks like it's too slow in practice. The benchmarks show decent numbers (in runs per second, ~300k insert, 60k delete, 100k search), but when used it's ~3x slower than the list-based version. I might just be doing something stupid, because it should be faster, but it isn't. 

Some remaining ideas: 

* A lot of intermediate list operations can be removed by hardcoding the number of children like [this haskell package](https://github.com/sebastian-philipp/r-tree/blob/master/Data/RTree/Base.hs#L83)
* linear splitting: faster in theory but for a small number of leaves/children the difference is probably negligible and the impact on search could be significant.

Also I'm not sure how your random insertion would work. Our input is sorted,performing a random shuffle seems costly. The quadratic split is much better (for delete and search) than the "cheapSplit" which evenly divides the children over two new nodes.  

[The full code is here if you're interested](https://gist.github.com/folkertdev/0d06872819fb115d167caf625fa10ccf)

**Comment by folkertdev** [comment]
I tried to get an overview of the uses of voronoi and delaunay in visualization and more broadly in computational geometry. Here is a bunch of notes/links. Much of it is probably not very relevant to elm right now but I thought it was helpful to get a wider perspective. 

actionable items:

* provide a way of associating a point with a voronoi cell and its vertex. 
    Like you suggested, we can use an r-tree here. 

* `centroid : Face -> (Direction2d, Maybe Float)`  
    The centroid is probably useful in general. In particular it can be used for finding the optimal direction and available space for a label. 
    For the one- and two-vertex faces we can only provide a direction, for 3-vertex faces a length is also computable.

* if the convex hull is going to be a separate data type, delaunay triangulation gives it for free.

most of this was taken from this excellent overview of voronoi/delaunay and d3 https://visionscarto.net/the-state-of-d3-voronoi


notes on voronoi: 

* use the vector from the data point to the centroid of the voronoi polygon.
    That gives a good indication of available space for a label, and the best direction of the label. 

    https://beta.observablehq.com/@mbostock/d3-voronoi-labels

* voronoi for getting the nearest point in a diagram is apparently useful because 
    we can use `mouseover` and `mouseout` on an svg path element with the polygon

    [d3 block showing this](https://bl.ocks.org/mbostock/8033015)

    But d3 now has a better method that works more generally, so you can get the nearest point without rendering to the dom
    However, for us the rtree-approach on data points seems much better, the d3 approach is not viable in elm 

    [link to D3 source](https://github.com/d3/d3-voronoi/blob/master/src/Diagram.js#L119)

* voronoi on different geometry 
    [This article](https://visionscarto.net/the-state-of-d3-voronoi
) shows voronoi on cylinder and torus (just the planar variant with the data duplicated on the edges), and sphere (mostly for map-related things). 

* weighted voronoi diagrams
    weighed 2d voronoi is a slice of unweighed 3d voronoi

* voronoi relaxation
    computed with Loyds algorithm: computes voronoi, computes the centroids of the polygons, and then repeates (so voronoi on the centroids, get new centroids, repeat) until stability.

    The relaxed voronoi can be used in stippling (I personally think this is super-interesting, but not so relevant to elm right now maybe)

    * [https://en.wikipedia.org/wiki/Stippling](https://en.wikipedia.org/wiki/Stippling)
    * [https://pdfs.semanticscholar.org/d881/4e859fb27a67dd13d54975924202bfaeaccc.pdf](https://pdfs.semanticscholar.org/d881/4e859fb27a67dd13d54975924202bfaeaccc.pdf)
    * [https://www.cs.ubc.ca/labs/imager/tr/2002/secord2002b/secord.2002b.pdf](https://www.cs.ubc.ca/labs/imager/tr/2002/secord2002b/secord.2002b.pdf)

    and mesh smoothing 

    * [http://www.math.ualberta.ca/ijnam/Volume-2-2005/Special/08.pdf](http://www.math.ualberta.ca/ijnam/Volume-2-2005/Special/08.pdf)
    

* use of non-euclidian distances 

* circular clipping 
    clip each polygon by a circle, instead of by edges of the bounding box. 


notes on delaunay triangulation 

* apparently in JS the delaunay triangulation of 10k points takes 31-150ms ([source](https://github.com/mapbox/delaunator#performance))

* Alpha shapes 


    remove edges longer than some length alpha ([wikipedia article](https://en.wikipedia.org/wiki/Alpha_shape))

    math-heavy paper on the topic: 
    https://graphics.stanford.edu/courses/cs268-11-spring/handouts/AlphaShapes/as_fisher.pdf
* related: beta skeleton 

    https://en.wikipedia.org/wiki/Beta_skeleton

* Urquhart graph
    Removes the longest edge of every triangle in the delaunay triangulation. The remaining graph is a good approximation of the relative neighborhood graph 
    (the graph connecting pairs of points p and q when there does not exist any third point r that is closer to both p and q than they are to each other)

    Can apparently be used to create nice polygons from the delaunay triangulation: 

    https://stackoverflow.com/questions/17926029/how-to-obtain-the-closed-polygons-from-the-relative-neighbor-graph

    And in general has uses in *computational morphology*, "the computational extraction of perceptually meaningful strucutre from dot patterns"

    http://rutcor.rutgers.edu/~dandrade/papers/ug.pdf

**Comment by ianmackenzie** [comment]
Ah, being able to trigger highlighting using `mouseover` and `mouseout` events etc. is an interesting point - I hadn't thought of that. Great resources in general! I'm particularly interested in mesh smoothing, since that looks like it might be a good approach to use for generating 'nice' meshes for things like finite element analysis (FEA). But seeing as it looks fairly straightforward to implement smoothing on top of the basic Voronoid diagram generation functionality I'm not sure there's a pressing need to add smoothing to `elm-geometry` itself.

You're correct, my C++/Scala versions don't include insertion or removal, since in most cases I only needed bulk creation.

I still think shuffling could be feasible - I actually took a stab recently at [a fast random shuffling function](https://github.com/ianmackenzie/elm-fast-shuffle-experiment/blob/master/Main.elm) based on tagging each list item with a randomly-generated value and then sorting based on that value, which should be pretty quick. But if that plus either your R-tree or a very simple (non-self-balancing) R-tree doesn't give a speedup, then I think it makes sense to split out the development of an efficient general-purpose R-tree as a separate issue and then update Delaunay triangulation to use it whenever it's done.

I think the next step is still to figure out an API for working with `DelaunayTriangulation2d` and `VoronoiDiagram2d` types - I didn't get a chance over the weekend but might be able to tonight or tomorrow. It might make sense to split the `Tesselation2d` modules into `DelaunayTriangulation2d` and `VoronoiDiagram2d` modules:

```elm
DelaunayTriangulation2d.fromPoints: List Point2d -> DelaunayTriangulation2d Point2d
DelaunayTriangulation2d.fromVerticesBy: (vertex -> Point2d) -> List vertex -> DelaunayTriangulation2d vertex
DelaunayTriangulation2d.toMesh : DelaunayTriangulation2d a -> TriangularMesh a
```

```elm
VoronoiDiagram2d.fromPoints : List Point2d -> VoronoiDiagram2d Point2d
VoronoiDiagram2d.fromVerticesBy : (vertex -> Point2d) -> List vertex -> VoronoiDiagram2d vertex
VoronoiDiagram2d.fromDelaunayTriangulation : DelaunayTriangulation2d vertex -> VoronoiDiagram2d vertex
```

**Comment by ianmackenzie** [comment]
Just finished an initial implementation of incremental Delaunay triangulation construction, along with Voronoi diagram construction - see what you think! The `DelaunayTriangulation` and `VoronoiRegions` examples under `media` now both allow dynamically modifying the result by moving the mouse over the image.

I think I'm mostly pretty happy with the API except for the return type of `DelaunayTriangulation2d.voronoiRegions` - not sure if those are the right field names, or even the right data structure in general. The initial idea was to have a separate `VoronoiDiagram2d` type/module, which could then have its own accessors, and that might be a better approach even if the entire API of that module is

```elm
fromDelaunayTriangulation : DelaunayTriangulation2d vertex -> VoronoiDiagram2d vertex
finiteRegions : VoronoiDiagram2d vertex -> List { vertex : vertex, region : Polygon2d }
infiniteRegions : VoronoiDiagram2d vertex -> List { vertex : vertex, leftAxis : Axis2d, rightAxis : Axis2d, finitePortion : Polyline2d }
```

Thoughts?

**Comment by folkertdev** [comment]
Should voronoi and delaunay really be so closely linked? When I just want voronoi, should I have to also learn about delaunay? The majority of users probably only wants voronoi diagrams. Providing some voronoi-specific wrappers around delaunay functions like 
```elm
VoronoiDiagram2d.fromVertices : (vertex -> Point2d) -> List vertex -> VoronoiDiagram2d vertex
```
looks nicer to me than having to refer to delaunay for creation and inserting/removing points.

For insertion specifically, a signature like `Dict.update` might be a good idea:
```elm
updateBy : (vertex -> Point2d) -> Point2d -> (Maybe vertex -> Maybe vertex) -> DelaunayTriangulation2d vertex -> DelaunayTriangulation2d vertex
```
it's important that the location can also be modified, so the `vertex -> Point2d` argument is required. 


Are left-axis and right-axis of the infinite regions really to the left and right of the polyline in global space? It seems most natural to have the first axis go through the first point of the polyline, and the second axis through the final one. Is it then guaranteed by the pseudo-angle ordering that the first is left and the second is right? 

It also looks like currently inserting a new vertex at an existing location will ignore the vertex, so it's impossible to update the other information besides position of a vertex.

**Comment by ianmackenzie** [comment]
The original plan was indeed to have `VoronoiDiagram2d` be separate so that you wouldn't have to think about Delaunay triangulations if you didn't care about them, but I struggled to figure out a Voronoi diagram API/implementation that would have obvious 'performance semantics'. For example, inserting a new point into a Voronoi diagram means an O(n) insertion into a Delaunay triangulation and then an O(n log n) full rebuild of the Voronoi diagram from the updated Delaunay triangulation. The current API makes it more clear that that's what happening, and I figured that most people who knew what a Voronoi diagram was would at least also know that it's related to the Delaunay triangulation and so might think to look in that module.

Do you think it's clear enough for a `VoronoiDiagram2d` module to basically mimic the `DelaunayTriangulation2d` API, where every function that constructs or modifies a `VoronoiDiagram2d` means performing the corresponding operation on an underlying `DelaunayTriangulation2d` and then rebuilding the Voronoi regions from scratch? I guess that does give us the opportunity to add optimizations in the future like recognizing which Delaunay triangles have been added/removed and only updating the relevant Voronoi regions...

I don't think an `updateBy` function is possible currently, since it would have to handle vertex *removal* and I don't know how to do that without rebuilding the entire triangulation. The Bowyer-Watson algorithm covers building a triangulation incrementally by adding vertices - I don't think it says anything about how to remove a vertex from an existing triangulation. If there is a way I would love to know it! The other issue is, as you mentioned, the current implementation ignores duplicate vertices. This means that if you inserted the same vertex twice and then removed it once, you'd end up with a triangulation without that vertex at all, which may or may not be what you want. (It's possible that the API should be more strict and return a `Result` from construction/insertion functions, returning `Err DuplicatePoints` if there's any duplication.)

The axes are 'left' and 'right' if you're at the vertex in question looking outwards to infinity (roughly speaking, left and right in an angular sense). These do in fact correspond to the start and end points of the polyline (the origin point of the left axis will be the start point of the polyline and the origin point of the right axis will be the end point of the polyline). Basically, every infinite Voronoi region has a 'U' shape; the `leftAxis` and `rightAxis` correspond to the left and right sides of the U, and the `finitePortion` corresponds to the curved bit at the bottom with the start point on the left and the end point on the right. I originally called the axes `startAxis` and `endAxis` but thought `startAxis` might sound like it was pointing inwards towards the start point instead of outwards from the start point towards infinity. I'm all for clearer names if we can think of them, but I also think a nice image in the documentation will make it pretty clear =)

**Comment by ianmackenzie** [comment]
Just updated from `elm-geometry` master - I have an idea for a reasonably clean way to trim infinite Voronoi regions to a bounding box so we can just expose

```elm
VoronoiDiagram2d.polygons : BoundingBox2d -> VoronoiDiagram2d vertex -> List ( vertex, Polygon2d )
```

but it required adding `LineSegment2d.intersectionWithAxis` (which I've been meaning to add anyways - see #5).

**Comment by ianmackenzie** [comment]
The proposed approach is based on the fact that the resulting trimmed polygon will always be convex, so it should be sufficient to just find all vertices of the resulting polygon (without attempting to keep track of order) and then sort them afterwards by pseudo-angle. I believe the set of vertices in the result polygon for a given infinite Voronoi region should be:

  - all polyline vertices inside the bounding box
  - all intersections with polyline edges and bounding box edges
  - all intersections with the left/right axes and bounding box edges
  - all bounding box vertices inside the Voronoi region (to the left of all polyline edges and both axes)

We'd also need to be able to trim *finite* Voronoi regions (polygons), in which case the logic would be

  - all polygon vertices inside the bounding box
  - all intersections with polygon edges and bounding box edges
  - all bounding box vertices inside the polygon (to the left of all polygon edges)

with a special optimization case of "if all polygon vertices are inside the bounding box, then just return the original polygon". If you can think of any cases where any of this would break down, let me know! I think the trickiest cases will be ones like axes that are exactly collinear with one side of a bounding box, but I think this can be handled by being non-strict about finding intersection points (e.g. return intersection points even if they occur right at the end of an edge instead of in the interior, accept vertices as inside the bounding box even if they're right on the edge), then add a final deduplication step after angular sorting.

**Comment by folkertdev** [comment]
Seems solid! 

line segments and axes that are colinear with and exactly on top of  one of the sides of the bounding box are tricky. I actually think it's fine to not count these as intersection points and only include the vertices, since the other rules will add the vertices of the bounding box when required. With this approach I don't think there would be duplicates at all. 

Anyway, we've seen that deduplication is actually pretty cheap so we can also err on the safe side by being lax with intersections and cleaning it up afterwards.

**Comment by ianmackenzie** [comment]
Just added an initial implementation of trimming Voronoi regions to a bounding box to get finite polygons - see what you think! I've updated the `media/src/VoronoiRegions.elm` example to use `VoronoiDiagram2d.polygons`.

**Comment by ianmackenzie** [comment]
Hey @gampleman I know you were interested in Delaunay triangulation and Voronoi diagram generation; could you take a look at the APIs of the `DelaunayTriangulation2d` and `VoronoiDiagram2d` modules in this PR and see what you think? I think this is getting close to complete, so it would be good to know if there are any major gaps in the current APIs or places where they are confusing. There are some examples of using both APIs in the `media/src` directory.

**Comment by ianmackenzie** [comment]
OK finally had time to finish the initial implementation of Voronoi region construction for collinear points - @folkertdev can you take a look and see if you can break it? Voronoi diagram construction should now return an `Err` on coincident vertices and return correct results for all valid cases (zero vertices, one vertex, two or more collinear vertices, or three or more non-collinear vertices).

**Comment by folkertdev** [comment]
It's looking very good. Can't see any obvious issues with it. I had to round the color rgb values to integers to make the colors show up. Possibly a chrome/svg issue. 

Also it is a bit ugly to have the `elm-geometry-svg` repo as a source-directory for `media`, but I guess it's the only way to make this work. Maybe this should be documented though; I'm not sure what the purpose of the media folder is exactly, is it just for developing, or actually a resource to point people to?

**Comment by ianmackenzie** [comment]
Good catch on the colors! In general I've been going back and forth between Chrome and Firefox, but I think that last bit I just tested on Firefox. Something to keep in mind for the future!

Yes, the `media` directory is meant to be purely internal - the intent was actually just to use that for Elm files used to generate images for documentation/release notes/the README etc. I abused it as an easy way to do some quick testing since I could copy-paste the existing polygon triangulation file. As an actual resource it would really be better to have an `examples` directory that just includes `elm-geometry` as a normal dependency - I'll have to do a bit of reorganization at some point.

Next steps for this PR are documentation/testing...I'm happy to handle documentation (I do kind of like to ensure that all documentation uses a consistent tone/style), and testing is kind of in limbo right now since I think `elm-test` for 0.19 might still have some Windows quirks and [`Fuzz.custom` is not yet supported](https://github.com/elm-explorations/test/issues/15), which I use for at least polygon fuzzing. That said, I think it should be possible to get testing mostly working by commenting out the polygon fuzzer and related tests, and then it should be possible to start adding some Delaunay triangulation/Voronoi region tests - @folkertdev is that something you might be able to take a stab at? Off the top of my head, would be good to do some fuzz testing for:

  - All Delaunay triangles have non-zero area
  - No Delaunay circumcircle contains any points other than the three points on it
  - If a Delaunay triangulation is empty, then all input points must be collinear
  - A Voronoi diagram should include a cell for every input vertex (as long as the clipping bounding box contains all input vertices)
  - Delaunay triangulation and Voronoi diagram construction should fail when coincident vertices are given
  - Every point inside a Voronoi region must be closer to the corresponding vertex than any other vertex

**Comment by folkertdev** [comment]
I can have a look, but likely we'll need another release of elm-test for that, because anything that combines fuzzers (internally implemented with fuzz2/andMap) can crash, see also [this issue](https://github.com/elm-explorations/test/issues/18)

**Comment by ianmackenzie** [comment]
Hmm...well, even getting the skeletons of some tests in (that at least compile) would be a good start. If you could take a look at starting that, I can try to start adding some documentation - we can take both of those to a certain point and then just merge this PR and I can finish up the details in `master` (e.g. as I do any necessary updating of the _rest_ of the tests to 0.19!).

**Comment by gampleman** [comment]
The code looks very clean, congratulations, this is very nice PR. I can maybe give some more feedback once I understand the public API better, perhaps once the docs are ready. I have some usecases in mind, so I’ll check them against those. I recently found a geographic voronoi implementation in terms of a regular Delaunay and a clever projection, I’m very excited to get that going as well.

**Comment by ianmackenzie** [comment]
Sounds good @gampleman! I'm glad to hear that about the geographic Voronoi diagram, hopefully that means `elm-geometry` can just stick to Cartesian coordinates for simplicity and not need to add dedicated geographic coordinates support. I should be able to work on documentation this weekend (long weekend woo!) so hopefully that will clarify things.

**Comment by folkertdev** [comment]
I pushed some basic skeletons for the tests but it's really inconvenient at the moment to write them. There is no editor support (for syntax mistakes etc.) because the compiler can't find the `Test` module, and I can't run the tests (and find out issues that way) because of the bug. Hopefully the next elm-test beta/release will fix this.

**Comment by ianmackenzie** [comment]
Just added some initial documentation - @gampleman hopefully this is enough to get you started, let me know if anything is confusing! Ultimately I'd like to add some nice images to better explain things, but there's a bit of infrastructure I'd like to build up there first (for example, I'm working on a little tool to conveniently take automated screenshots of Elm pages, to make it really easy to generate doc example images...)

**Comment by ianmackenzie** [comment]
OK, I'm going to go ahead and merge this - @gampleman, if you have any feedback, I would say still leave it as a comment here and I'll just make any necessary changes in `master`. (I'd like to update `elm-geometry` to use `elm-format` 0.8, but that will cause a bunch of code churn so will be easiest to do if there are no open pull requests that will need to be merged later.)

File: src/VoronoiDiagram2d.elm [hunk 1]
```diff
@@ -0,0 +1,866 @@
...
+
+
+# Modification
+
+@docs insertPoint, insertVertexBy
+
+
+# Properties
+
+@docs vertices, polygons
```
[the remark this answers is not in the document]
Comment (ianmackenzie) [reply src/VoronoiDiagram2d.elm]: Just added #69 to track - I don't see an easier way currently than just post-processing the polygons, though (that is, off the top of my head I don't think there's any part of the internal data structure we can exploit to make that easier/faster).

**Comment by ianmackenzie** [comment]
Good points @gampleman! I've added a bit of introductory documentation to each module as well as some complexity notes.

**Pull request merged**